### PR TITLE
Add a new module: ansible_playbook

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY
This pull request adds a new module to run an Ansible playbook on the managed host. The rational is that I have faced many situations where I had to deploy playbooks on the managed host (from an archive or a SCM) and then run it from this host, using a generated inventory. Most of the time I run a predefined command, so that `command` module is enough, but recently I had to provide a generic way to run playbooks from jump hosts, using a playbook that setups the jump host and then runs the playbook. Building the command line becomes cumbersome if we have to handle all the `ansible-playbook` options. This module is mainly a wrapper and is based on the `command` module.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
ansible_playbook

##### ANSIBLE VERSION
```
ansible 2.5.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/fdupont/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

##### ADDITIONAL INFORMATION
This is the first module I contribute, so any guidance is appreciated.

### CODE REVIEW

**Comment by webknjaz** [comment]
Jumphost? Sounds like a use-case for `ssh -A`

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (webknjaz) [inline lib/ansible/modules/remote_management/ansible_playbook.py]: Did you write it 6 years ago?

Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: Nice catch @webknjaz. That's what happens with copy/paste.

**Comment by ghost** [comment]
@webknjaz 
Jumphost is part of the use case. Here, I have to run the playbook from a different host. It is similar to https://docs.ansible.com/ansible-tower/latest/html/administration/clustering.html#isolated-instance-groups.

**Comment by bcoca** [comment]
why not use 'delegate_to' instead?

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/ansible_playbook.py]: Instead of listing here, point at `ansible-doc -t conneciton -l` and maybe link to the connection plugins page, since this can change by installation.

**Comment by ghost** [comment]
@bcoca 
Using `delegate_to` doesn't change you execution context. You still have the same inventory, same facts, same SSH options, etc... unless you override them at the task/block level. This can become very complex to implement as options of the task/block, while using this module allows you to run a playbook in a brand new context: new inventory, SSH options, tags...

An example is OpenShift deployment. You create new VMs using a playbook. One of these VMs is the deployment host. So, your playbook can then look like:

1. Create all your VMs
2. Generate your OpenShift deployment inventory based on a template
3. Run the deployment playbook (provided by an RPM or from GitHub) on the deployment host using this generated inventory
4. Wait for the playbook to complete
5. Perform post configuration tasks

Here, you are sure that your playbook execution context is clean.

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: That would be the case for all options as it is a wrapper for ansible-playbook, wouldn't it ?

**Comment by bcoca** [comment]
Then you can just the ssh args variables to define a proxy, you don't need to do it yourself in the module in a 'non standard' fashion. Both cases are already covered, with 'current context' (delegate to) and with 'remote context' (ssh args settings).

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (bcoca) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: yep, for all 'plugin based options'

**Comment by ghost** [comment]
@bcoca 
So you mean, building the `ansible-playbook` command as a variable in the initial context, then using `command` module with `delegate_to` ? The problem then is to build the command line using tasks: it requires a lot of tasks to cover all the options, and manipulating hashes and arrays in tasks is not that easy. The module handles all that with a single task.

**Comment by bcoca** [comment]
no, no mention of `command` module, unsure what you are going into here. 

I can see 2 ways you need 'proxies' either one to execute THIS module on, which `delegate_to` and ssh args in 'local' config/inventory handles already, another is to execute the tasks in the playbook itself, which are also covered by either `delegate_to` or ssh args in remote setting/inventory an ssh proxy.

**Comment by ghost** [comment]
@bcoca 
I'm a bit lost and confused. Would you mind elaborating your use of `delegate_to` and ssh args in running a playbook on a remote host with an example ?

**Comment by bcoca** [comment]
Not sure what the confusion is, its the same feature that already exists, I'm trying to find a scenario in which this is not sufficient and requires an additional proxy, but here are a couple of examlpes anyways:

```
- name: delegated
  ansible_playbook:  ...
  delegate_to: myexecutor

- name: with proxy
  ansible_playbook:  ...
  vars:
     ansible_ssh_common_args: ' -o ProxyCommand="ssh -W %h:%p -q user@myjumphosttotargetexecutor"'
```

These are just from the controller's view, you can do both again on the 'executor' and even chain/mix match ... it gets crazy, but i don't see a need for another proxy.

**Comment by ghost** [comment]
@bcoca 
Thanks for examples. But I don't see where I define another proxy.

From the beginning, `delegate_to` is expected to be used, as there is a single *executor*. The inventory is generated on `myexecutor` using `copy`, `template` or any other file manipulation module, then the playbook is run using `ansible_playbook` module and `delegate_to: myexecutor`.

I have implemented (almost) all the `ansible-playbook` options as parameters, in order to give the user as much control as possible on how the playbook is run. But I don't expect they will all be used.

**Comment by bcoca** [comment]
that does not clarify to me why you need the proxy setting, you can just add it to the vars or cli on the executor (inventory/extras/etc)

**Comment by ghost** [comment]
What proxy setting ? I think the misunderstanding is with the term *proxy*.

**Comment by bcoca** [comment]
nvmd ...i think im just going (more) nuts ... i could swear you had a `proxy` option to the module but now i cannot find it ... ignore me ... men in white coats arriving soon ...

**Comment by ghost** [comment]
@bcoca No need for men in white coats ;) I think the initial comments about jump host were a bit confusing.

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: Is there a way to generate the choices list ? As you say, it may change over time, so having a dynamic list would ease the maintenance.

Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: @bcoca 
Thinking about it, the problem with linking to the documentation without providing the actual list is that the CI will complain that implementation and documentation are not consistent. Should I just add the link to the documentation and keep the list ?

**Comment by bcoca** [comment]
i would not hardcode the list in either spot, its as bad in argspec as in docs, that would prevent users from using new/custom plugins

**Comment by ghost** [comment]
Then how would you implement that ? Can I get the list of plugins and their arguments to build both lists ? I don't expect you to do it, just a pointer to start :smile:

**Comment by bcoca** [comment]
I would not bother, because you can easily get a list for the 'controller' .. but you need it from the remote that actually executes it, the lists might be different depending on what is available.

I would document that its the list of plugins available to that Ansible installation and then deal with the errors of 'incorrect plugin names' at execution.

**Comment by ghost** [comment]
If I understand the code correctly, the only option based on a plugin is `connection`. Am I right ?
I have updated the documentation and argspecs to link to the Ansible documentation. Is it enough ?

**Comment by ghost** [comment]
@bcoca Is there anything else I have to do on this PR ?

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/ansible_playbook.py]: use /latest/ instead of version specific, since we will eventually archive docs for older versions

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (bcoca) [inline lib/ansible/modules/remote_management/ansible_playbook.py:95]: missing 'machinectl'

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py]: Thanks. Done.

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (ghost) [reply lib/ansible/modules/remote_management/ansible_playbook.py:95]: Thanks. Done.

**Comment by bcoca** [comment]
@fdupont-redhat i brought this up in recent IRC meeting, there are several issues to address, from 'is this a feature we want in core?'  to issues with the implementation. I've added this to the agenda for the next meeting, you can attend to make your case there.

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/ansible_playbook.py:43]: If it is not required, you don't have to state this explicitly. (Makes it easier to see which parameters are required when skimming the list)

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/ansible_playbook.py:81]: Please make this:
```yaml
    default: 'no'
```
as the default list of choices for booleans are `[ 'no', 'yes' ]`.

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (dagwieers) [reply lib/ansible/modules/remote_management/ansible_playbook.py:95]: The problem here is that it has to be maintained outside of Ansible.
Might be better to leave it as free-form (and make the module fail message give details).

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (dagwieers) [inline lib/ansible/modules/remote_management/ansible_playbook.py:29]: All descriptions (except the short_description) end with a trailing dot.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
Additional functionality:
- Add a 'verbosity' parameter which turns 0-9 into -q, -v, -vv, -vvv, ...
- Run with JSON output and return a JSON blob that can be registered and used
- Pass task check-mode and diff-mode to ansible-playbook run

**Comment by abadger** [comment]
At today's meeting we discussed this and there was a general reluctance to merge this into the core.  There were a few factors in that which could be mitigated.  However, it's not certain if the mitigations would be enough to gain support for adding this.  Here's our thinking:

* Confusion factor
  * We don't believe we, the core team, would want to officially support this so it would remain a community supported module.  However, being named ansible_playbook and dealing with running ansible would inevitably bring some support burden onto us.
    * Renaming the module could mitigate this to some extent.  unofficial_playbook_runner was jokingly put out there.  We didn't have a great, serious idea for a name
    * Add documentation that the ansible core team doesn't support this and alternatives exist (ansible-pull? ansible-runner?)
  * Complications can still arise just by ansible invoking ansible... core team would have to figure out if a problem was inherent to ansible or due to the way the module ran ansible.

* Does this module add some value over just using command?
  * In the past we've looked at whether a module is idempotent or adds checkmode as two quantifiable indications of value add.  We noticed that this module could add those easily but at present it isn't adding either (it says supportscheckmode=True in the argspec but doesn't add the --check cli arg based upon it.  It also unconditionally returns changed=True instead of looking at whether the remotely invoked ansible succeeded or failed.
  * This is currently just returning the stdout and stderr output back to ansible.  It seems like using the json callback plugin and returning a json data structure would be much more flexible

* Is this the right way to implement this?
  * In the past, we've told people to use ansible-pull and awx/tower for some of these things.  In the present, there's the ansible-runner project which can also handle some of this.  Perhaps a module implemented via ansible-runner would be better?

* We wanted to gather more information...
  * After talking to bcoca, it seems that there isn't presently a way to do precisely what you are doing in a good way.  However, it's unclear if this is the way forward or not.  I've asked Jesse Keating who has a long-term interest in this sort of thing to give some thoughts on this as an external view of what about this might be good and what might be better done some other way.

I want to sum up by saying this isn't a "no, we don't want this" but it's also far from a "yes, we'll definitely take this once these issues are fixed".  There's a lot of thought we want to put into the ramifications of this before making a decision.  In the meantime, we'll try to give helpful feedback on making it better and even if we decide we don't want to merge this, hopefully the feedback you get would make it a better module that you could distribute via galaxy.ansible.com.

Thanks!

File: lib/ansible/modules/remote_management/ansible_playbook.py
Comment (SpamapS) [inline lib/ansible/modules/remote_management/ansible_playbook.py:392]: Shouldn't this be `--forks` ?

**Review by SpamapS — COMMENTED** [review]
IMO this doesn't add enough value over the `command:` module to justify the module.  If it were to copy the playbook up at least, or build inventory from structured data, or validate the args before trying to execute, yes. But IMO, I'd rather just build an ansible-playbook command line using jinja than maintain a big chunk of python.

**Comment by ghost** [comment]
@abadger Thanks for that long and educative explanation. It gave me the feeling that creating a module that leverages [Ansible Runner Python interface](https://github.com/ansible/ansible-runner/blob/master/docs/python_interface.rst) would be a better idea. I reached out to @matburt and he's _"definitely down for a module that directly interfaces with Runner itself"_. I'll try to outline a solution based on that, making it a fourth way to invoke Ansible Runner. How does it sound ?

**Comment by abadger** [comment]
That sounds good!

**Comment by ant31** [comment]
I've just created the exact same feature request (failing to find this one): #45596  

I'm going to details my use case, just in case it helps to value this module. 
We are building a ci infra to continuously test our ansible-scripts ([kubernetes/kubespray](https://github.com/kubernetes-incubator/kubespray)) before merging PRs. 

- We have N hosts to start VMs  
    VMs are not accessible outside the host. 

- Our CI launch a job that: 
    - Creates VMs on different selected hosts
    - Generates a unique inventory with the vm ips on each
    - Starts an ansible-playbook (must be run from the hosts)

For now, we are using `command`, we could probably  use bastion with parameterized ssh_args on those tasks (to select which host to jump to), but an ansible_playbook module seems a much more natural and canonical way to invoke commands using ansible.  

```
- name: generate inventory and group_vars
  template: 
     src: inventory.yaml.j2
     
- name: Deploy cluster
  command: ansible-playbook cluster.yml -i {{build_dir}}/inventory.yaml
  register: status
```

**Comment by jhg03a** [comment]
I do this today with the command module, just for the ability to async playbook execution.

**Pull request closed without merging**
